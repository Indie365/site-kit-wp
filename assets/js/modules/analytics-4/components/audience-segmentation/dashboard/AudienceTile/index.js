/**
 * AudienceTile component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import InfoTooltip from '../../../../../../components/InfoTooltip';
import ChangeBadge from '../../../../../../components/ChangeBadge';
import AudienceMetricIconVisitors from '../../../../../../../svg/icons/audience-metric-icon-visitors.svg';
import AudienceMetricIconVisitsPerVisitor from '../../../../../../../svg/icons/audience-metric-icon-visits-per-visitor.svg';
import AudienceMetricIconPagesPerVisit from '../../../../../../../svg/icons/audience-metric-icon-pages-per-visit.svg';
import AudienceMetricIconPageviews from '../../../../../../../svg/icons/audience-metric-icon-pageviews.svg';
import AudienceMetricIconCities from '../../../../../../../svg/icons/audience-metric-icon-cities.svg';
import AudienceMetricIconTopContent from '../../../../../../../svg/icons/audience-metric-icon-top-content.svg';
import {
	BREAKPOINT_SMALL,
	BREAKPOINT_TABLET,
	useBreakpoint,
} from '../../../../../../hooks/useBreakpoint';
import AudienceTileMetric from './AudienceTileMetric';
import AudienceTileCitiesMetric from './AudienceTileCitiesMetric';
import AudienceTilePagesMetric from './AudienceTilePagesMetric';
import { numFmt } from '../../../../../../util';

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';

// TODO: as part of #8484 the report props should be updated to expect
// the full report rows for the current tile to reduce data manipulation
// in AudienceTiles.
export default function AudienceTile( {
	title,
	infoTooltip,
	visitors,
	visitsPerVisitor,
	pagesPerVisit,
	pageviews,
	percentageOfTotalPageViews,
	topCities,
	topContent,
	topContentTitles,
	Widget,
} ) {
	const breakpoint = useBreakpoint();

	return (
		<Widget noPadding>
			<div className="googlesitekit-audience-segmentation-tile">
				{ breakpoint !== BREAKPOINT_TABLET &&
					breakpoint !== BREAKPOINT_SMALL && (
						<div className="googlesitekit-audience-segmentation-tile__header">
							<div className="googlesitekit-audience-segmentation-tile__header-title">
								{ title }
								{ infoTooltip && (
									<InfoTooltip
										title={ infoTooltip }
										tooltipClassName="googlesitekit-info-tooltip__content--audience"
									/>
								) }
							</div>
						</div>
					) }
				<div className="googlesitekit-audience-segmentation-tile__metrics">
					<AudienceTileMetric
						TileIcon={ AudienceMetricIconVisitors }
						title={ __( 'Visitors', 'google-site-kit' ) }
						metricValue={ visitors.currentValue }
						Badge={ () => (
							<ChangeBadge
								previousValue={ visitors.previousValue }
								currentValue={ visitors.currentValue }
							/>
						) }
					/>

					<AudienceTileMetric
						TileIcon={ AudienceMetricIconVisitsPerVisitor }
						title={ __( 'Visits per visitor', 'google-site-kit' ) }
						metricValue={ visitsPerVisitor.currentValue }
						Badge={ () => (
							<ChangeBadge
								previousValue={ visitsPerVisitor.previousValue }
								currentValue={ visitsPerVisitor.currentValue }
							/>
						) }
					/>

					<AudienceTileMetric
						TileIcon={ AudienceMetricIconPagesPerVisit }
						title={ __( 'Pages per visit', 'google-site-kit' ) }
						metricValue={ pagesPerVisit.currentValue }
						Badge={ () => (
							<ChangeBadge
								previousValue={ pagesPerVisit.previousValue }
								currentValue={ pagesPerVisit.currentValue }
							/>
						) }
					/>

					<AudienceTileMetric
						TileIcon={ AudienceMetricIconPageviews }
						title={ sprintf(
							/* translators: %s: is a percentage value such as 33.3%. */
							__( '%s of total pageviews', 'google-site-kit' ),
							numFmt( percentageOfTotalPageViews, {
								style: 'percent',
								maximumFractionDigits: 1,
							} )
						) }
						metricValue={ pageviews.currentValue }
						Badge={ () => (
							<ChangeBadge
								previousValue={ pageviews.previousValue }
								currentValue={ pageviews.currentValue }
							/>
						) }
					/>

					<AudienceTileCitiesMetric
						TileIcon={ AudienceMetricIconCities }
						title={ __(
							'Cities with the most visitors',
							'google-site-kit'
						) }
						topCities={ topCities }
					/>

					<AudienceTilePagesMetric
						TileIcon={ AudienceMetricIconTopContent }
						title={ __(
							'Top content by pageviews',
							'google-site-kit'
						) }
						topContentTitles={ topContentTitles }
						topContent={ topContent }
					/>
				</div>
			</div>
		</Widget>
	);
}

AudienceTile.propTypes = {
	title: PropTypes.string.isRequired,
	infoTooltip: PropTypes.oneOfType( [ PropTypes.string, PropTypes.element ] ),
	visitors: PropTypes.object,
	visitsPerVisitor: PropTypes.object,
	pagesPerVisit: PropTypes.object,
	pageviews: PropTypes.object,
	percentageOfTotalPageViews: PropTypes.number,
	topCities: PropTypes.object,
	topContent: PropTypes.object,
	topContentTitles: PropTypes.object,
	Widget: PropTypes.elementType.isRequired,
};
