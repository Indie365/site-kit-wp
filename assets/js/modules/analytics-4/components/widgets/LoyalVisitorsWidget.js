/**
 * LoyalVisitorsWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import {
	MetricTileNumeric,
	whenKeyMetricsWidgetVisible,
} from '../../../../components/KeyMetrics';
import { pickIntValueWhere } from '../../utils';
import { numFmt } from '../../../../util';

const { useSelect, useInViewSelect } = Data;

function LoyalVisitorsWidget( { Widget } ) {
	const dates = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
			compare: true,
		} )
	);

	const reportOptions = {
		...dates,
		dimensions: [ 'newVsReturning' ],
		metrics: [ { name: 'activeUsers' } ],
	};

	const report = useInViewSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getReport( reportOptions )
	);

	const loading = useInViewSelect(
		( select ) =>
			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
				'getReport',
				[ reportOptions ]
			)
	);

	const { rows = [], totals = [] } = report || {};

	// Total users and returning users for the current date range.
	const returning = pickIntValueWhere( rows, 'metricValues.0.value', {
		'dimensionValues.0.value': 'returning',
		'dimensionValues.1.value': 'current_range',
	} );
	const total = pickIntValueWhere( totals, 'metricValues.0.value', {
		'dimensionValues.1.value': 'current_range',
	} );

	// Total users and returning users for the previous date range.
	const prevReturning = pickIntValueWhere( rows, 'metricValues.0.value', {
		'dimensionValues.0.value': 'returning',
		'dimensionValues.1.value': 'compare_range',
	} );
	const prevTotal = pickIntValueWhere( totals, 'metricValues.0.value', {
		'dimensionValues.1.value': 'compare_range',
	} );

	// Calculate portions.
	const currentPercentage = total > 0 ? returning / total : 0;
	const prevPercentage = prevTotal > 0 ? prevReturning / prevTotal : 0;

	const format = {
		style: 'percent',
		signDisplay: 'never',
		maximumFractionDigits: 1,
	};

	return (
		<MetricTileNumeric
			Widget={ Widget }
			title={ __( 'Loyal visitors', 'google-site-kit' ) }
			metricValue={ currentPercentage }
			metricValueFormat={ format }
			subText={ sprintf(
				/* translators: %d: Number of total visitors visiting the site. */
				__( 'of %d total visitors', 'google-site-kit' ),
				total
			) }
			previousValue={ prevPercentage }
			currentValue={ currentPercentage }
			loading={ loading }
			tooltip={ sprintf(
				// translators: %1$s - the previous value, %2$s - the current value
				__( '%1$s before vs %2$s now', 'google-site-kit' ),
				numFmt( prevPercentage, format ),
				numFmt( currentPercentage, format )
			) }
		/>
	);
}

LoyalVisitorsWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
};

export default whenKeyMetricsWidgetVisible()( LoyalVisitorsWidget );
