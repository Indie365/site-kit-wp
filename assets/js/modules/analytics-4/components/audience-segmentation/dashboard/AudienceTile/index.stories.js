/**
 * AudienceTile Component Stories.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import WithRegistrySetup from '../../../../../../../../tests/js/WithRegistrySetup';
import { Provider as ViewContextProvider } from '../../../../../../components/Root/ViewContextContext';
import {
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
} from '../../../../../../googlesitekit/constants';
import { withWidgetComponentProps } from '../../../../../../googlesitekit/widgets/util';
import AudienceTile from '.';

const WidgetWithComponentProps =
	withWidgetComponentProps( 'audienceTile' )( AudienceTile );

function Template( { setupRegistry = () => {}, viewContext, ...args } ) {
	return (
		<WithRegistrySetup func={ setupRegistry }>
			<ViewContextProvider
				value={ viewContext || VIEW_CONTEXT_MAIN_DASHBOARD }
			>
				<WidgetWithComponentProps { ...args } />
			</ViewContextProvider>
		</WithRegistrySetup>
	);
}

// TODO: As part of #8484, update these stories to use the data-mock
// functions to provide report data rather than hardcoding props.
const readyProps = {
	title: 'New visitors',
	toolTip: 'This is a tooltip',
	visitors: {
		metricValue: 24200,
		currentValue: 24200,
		previousValue: 20424,
	},
	visitsPerVisitor: {
		metricValue: 3,
		currentValue: 3,
		previousValue: 2,
	},
	pagesPerVisit: {
		metricValue: 2,
		currentValue: 2,
		previousValue: 3,
	},
	pageviews: {
		metricValue: 1565,
		currentValue: 1565,
		previousValue: 1504,
	},
	percentageOfTotalPageViews: 0.333,
	topCities: {
		dimensionValues: [
			{
				value: 'Dublin',
			},
			{
				value: 'London',
			},
			{
				value: 'New York',
			},
		],
		metricValues: [
			{
				value: 0.388,
			},
			{
				value: 0.126,
			},
			{
				value: 0.094,
			},
		],
		total: 0.608,
	},
	topContent: {
		dimensionValues: [
			{
				value: '/en/test-post-1/',
			},
			{
				value: '/en/test-post-2/',
			},
			{
				value: '/en/test-post-3/',
			},
		],
		metricValues: [
			{
				value: 847,
			},
			{
				value: 596,
			},
			{
				value: 325,
			},
		],
		total: 1768,
	},
	topContentTitles: {
		'/en/test-post-1/': 'Test Post 1',
		'/en/test-post-2/': 'Test Post 2',
		'/en/test-post-3/': 'Test Post 3',
	},
};

export const Ready = Template.bind( {} );
Ready.storyName = 'Ready';
Ready.args = readyProps;
Ready.scenario = {
	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTile/Ready',
};

export const ReadyWithToolTip = Template.bind( {} );
ReadyWithToolTip.storyName = 'ReadyWithToolTip';
ReadyWithToolTip.args = {
	...readyProps,
	infoTooltip: 'This is a tooltip',
};
ReadyWithToolTip.scenario = {
	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTile/ReadyWithToolTip',
};

export const ReadyViewOnly = Template.bind( {} );
ReadyViewOnly.storyName = 'ReadyViewOnly';
ReadyViewOnly.args = {
	...readyProps,
	viewContext: VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
};
ReadyViewOnly.scenario = {
	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTile/ReadyViewOnly',
};

export const ReadyLongCityNames = Template.bind( {} );
ReadyLongCityNames.storyName = 'ReadyLongCityNames';
ReadyLongCityNames.args = {
	...readyProps,
	topCities: {
		dimensionValues: [
			{
				value: 'Llanfairpwllgwyngyllgogerychwyrndrobwllllantysiliogogogoch',
			},
			{
				value: 'Lake Chaubunagungamaug',
			},
			{
				value: 'Taumatawhakatangihangakoauauotamateaturipukakapikimaungahoronukupokaiwhenuakitanatahu',
			},
		],
		metricValues: [
			{
				value: 0.388,
			},
			{
				value: 0.126,
			},
			{
				value: 0.094,
			},
		],
		total: 0.608,
	},
};
ReadyLongCityNames.scenario = {
	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTile/ReadyLongCityNames',
};

export const NoData = Template.bind( {} );
NoData.storyName = 'NoData';
NoData.args = {
	...readyProps,
	topCities: null,
	topContent: null,
};
NoData.scenario = {
	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTile/NoData',
};

export default {
	title: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceTile',
};
