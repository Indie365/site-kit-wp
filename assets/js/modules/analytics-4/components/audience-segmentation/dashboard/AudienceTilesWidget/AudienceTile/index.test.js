/**
 * AudienceTile component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { useIntersection as mockUseIntersection } from 'react-use';

/**
 * Internal dependencies
 */
import AudienceTile from '.';
import {
	act,
	fireEvent,
	render,
} from '../../../../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideModuleRegistrations,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
	waitForDefaultTimeouts,
} from '../../../../../../../../../tests/js/utils';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '../../../../../../../googlesitekit/constants';
import { CORE_SITE } from '../../../../../../../googlesitekit/datastore/site/constants';
import { CORE_USER } from '../../../../../../../googlesitekit/datastore/user/constants';
import { withWidgetComponentProps } from '../../../../../../../googlesitekit/widgets/util';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '../../../../../../../util/errors';
import * as tracking from '../../../../../../../util/tracking';
import {
	MODULES_ANALYTICS_4,
	DATE_RANGE_OFFSET,
} from '../../../../../datastore/constants';
import { provideCustomDimensionError } from '../../../../../utils/custom-dimensions';
import { getAnalytics4MockResponse } from '../../../../../utils/data-mock';

jest.mock( 'react-use', () => ( {
	...jest.requireActual( 'react-use' ),
	useIntersection: jest.fn(),
} ) );

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'AudienceTile', () => {
	let registry;

	const WidgetWithComponentProps =
		withWidgetComponentProps( 'audienceTile' )( AudienceTile );

	const audienceResourceName = 'properties/12345/audiences/12345';
	const props = {
		audienceResourceName,
		title: 'New visitors',
		toolTip: 'This is a tooltip',
		loaded: true,
		visitors: {
			metricValue: 24200,
			currentValue: 24200,
			previousValue: 20424,
		},
		visitsPerVisitor: {
			metricValue: 3,
			currentValue: 3,
			previousValue: 2,
		},
		pagesPerVisit: {
			metricValue: 2,
			currentValue: 2,
			previousValue: 3,
		},
		pageviews: {
			metricValue: 1565,
			currentValue: 1565,
			previousValue: 1504,
		},
		percentageOfTotalPageViews: 0.333,
		topCities: {
			dimensionValues: [
				{
					value: 'Dublin',
				},
				{
					value: 'London',
				},
				{
					value: 'New York',
				},
			],
			metricValues: [
				{
					value: 0.388,
				},
				{
					value: 0.126,
				},
				{
					value: 0.094,
				},
			],
			total: 0.608,
		},
		topContent: {
			dimensionValues: [
				{
					value: '/en/test-post-1/',
				},
				{
					value: '/en/test-post-2/',
				},
				{
					value: '/en/test-post-3/',
				},
			],
			metricValues: [
				{
					value: 847,
				},
				{
					value: 596,
				},
				{
					value: 325,
				},
			],
			total: 1768,
		},
		topContentTitles: {
			'/en/test-post-1/': 'Test Post 1',
		},
		isZeroData: false,
		isPartialData: false,
	};

	beforeEach( () => {
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: false,
			intersectionRatio: 0,
		} ) );

		registry = createTestRegistry();
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
		] );
		provideModuleRegistrations( registry );
		provideUserAuthentication( registry );

		registry.dispatch( CORE_USER ).setReferenceDate( '2021-01-28' );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			availableCustomDimensions: [],
			availableAudiencesLastSyncedAt: ( Date.now() - 1000 ) / 1000,
		} );

		const dates = registry.select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
			compare: true,
		} );

		const options = {
			...dates,
			dimensions: [ 'date' ],
			metrics: [
				{
					name: 'totalUsers',
				},
			],
		};
		const reportData = getAnalytics4MockResponse( options );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetReport( reportData, {
			options,
		} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.finishResolution( 'getReport', [ options ] );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveIsGatheringData( false );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: {
					[ audienceResourceName ]: 20201220,
				},
				customDimension: {},
				property: {},
			} );
	} );

	afterEach( () => {
		mockTrackEvent.mockClear();
	} );

	it( 'should render the AudienceTile component', () => {
		const { container } = render(
			<WidgetWithComponentProps { ...props } />,
			{
				registry,
			}
		);

		expect( container ).toMatchSnapshot();
	} );

	describe( 'Top content metric', () => {
		it( 'should track an event when the create custom dimension CTA is viewed', () => {
			const { getByRole, rerender } = render(
				<WidgetWithComponentProps { ...props } />,
				{
					registry,
					viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
				}
			);

			expect(
				getByRole( 'button', { name: /update/i } )
			).toBeInTheDocument();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			// Simulate the CTA becoming visible.
			mockUseIntersection.mockImplementation( () => ( {
				isIntersecting: true,
				intersectionRatio: 1,
			} ) );

			rerender( <WidgetWithComponentProps { ...props } /> );

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-top-content-cta',
				'view_cta'
			);
		} );

		it( 'should track an event when the create custom dimension CTA is clicked', async () => {
			const { getByRole } = render(
				<WidgetWithComponentProps { ...props } />,
				{
					registry,
					viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
				}
			);

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			await act( async () => {
				fireEvent.click( getByRole( 'button', { name: /update/i } ) );

				// Allow the `trackEvent()` promise to resolve.
				await waitForDefaultTimeouts();
			} );

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-top-content-cta',
				'create_custom_dimension'
			);
		} );
	} );

	describe( 'AudienceErrorModal', () => {
		it( 'should show the OAuth error modal when the required scopes are not granted', async () => {
			provideSiteInfo( registry, {
				setupErrorCode: 'access_denied',
			} );

			provideUserAuthentication( registry, {
				grantedScopes: [],
			} );

			const { getByRole, getByText } = render(
				<WidgetWithComponentProps { ...props } />,
				{
					registry,
				}
			);

			expect(
				getByRole( 'button', { name: /update/i } )
			).toBeInTheDocument();

			await act( async () => {
				fireEvent.click( getByRole( 'button', { name: /update/i } ) );

				// Allow the `trackEvent()` promise to resolve so the custom dimension creation logic can be executed.
				await waitForDefaultTimeouts();
			} );

			// Verify the error is an OAuth error variant.
			expect(
				getByText( /Analytics update failed/i )
			).toBeInTheDocument();

			// Verify the "Get help" link is displayed.
			expect( getByText( /get help/i ) ).toBeInTheDocument();

			expect(
				getByRole( 'link', { name: /get help/i } )
			).toHaveAttribute(
				'href',
				registry.select( CORE_SITE ).getErrorTroubleshootingLinkURL( {
					code: 'access_denied',
				} )
			);

			// Verify the "Retry" button is displayed.
			expect( getByText( /retry/i ) ).toBeInTheDocument();
		} );

		it( 'should show the insufficient permission error modal when the user does not have the required permissions', async () => {
			const error = {
				code: 'test_error',
				message: 'Error message.',
				data: { reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS },
			};

			provideCustomDimensionError( registry, {
				customDimension: 'googlesitekit_post_type',
				error,
			} );

			const { getByRole, getByText } = render(
				<WidgetWithComponentProps { ...props } />,
				{
					registry,
				}
			);

			expect(
				getByRole( 'button', { name: /update/i } )
			).toBeInTheDocument();

			await act( async () => {
				fireEvent.click( getByRole( 'button', { name: /update/i } ) );

				// Allow the `trackEvent()` promise to resolve so the custom dimension creation logic can be executed.
				await waitForDefaultTimeouts();
			} );

			// Verify the error is "Insufficient permissions" variant.
			expect(
				getByText( /Insufficient permissions/i )
			).toBeInTheDocument();

			// Verify the "Get help" link is displayed.
			expect( getByText( /get help/i ) ).toBeInTheDocument();

			// Verify the "Request access" button is displayed.
			expect( getByText( /request access/i ) ).toBeInTheDocument();
		} );

		it( 'should show the generic error modal when an internal server error occurs', async () => {
			const error = {
				code: 'internal_server_error',
				message: 'Internal server error',
				data: { status: 500 },
			};

			provideCustomDimensionError( registry, {
				customDimension: 'googlesitekit_post_type',
				error,
			} );

			const { getByRole, getByText } = render(
				<WidgetWithComponentProps { ...props } />,
				{
					registry,
				}
			);

			expect(
				getByRole( 'button', { name: /update/i } )
			).toBeInTheDocument();

			await act( async () => {
				fireEvent.click( getByRole( 'button', { name: /update/i } ) );

				// Allow the `trackEvent()` promise to resolve so the custom dimension creation logic can be executed.
				await waitForDefaultTimeouts();
			} );

			// Verify the error is a generic error variant.
			expect(
				getByText( /Failed to enable metric/i )
			).toBeInTheDocument();

			// Verify the "Retry" button is displayed.
			expect(
				getByRole( 'button', { name: /retry/i } )
			).toBeInTheDocument();
		} );
	} );
} );
