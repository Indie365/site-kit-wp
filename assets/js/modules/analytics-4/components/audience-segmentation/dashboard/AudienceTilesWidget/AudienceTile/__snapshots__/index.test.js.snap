// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudienceTile should render the AudienceTile component 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--audienceTile googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-audience-segmentation-tile"
      >
        <div
          class="googlesitekit-audience-segmentation-tile__metrics"
        >
          <div
            class="googlesitekit-audience-segmentation-tile-metric"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-metric__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__container"
            >
              <div
                class="googlesitekit-audience-segmentation-tile-metric__value"
              >
                24K
              </div>
              <div
                class="googlesitekit-audience-segmentation-tile-metric__title"
              >
                Visitors
              </div>
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__badge-container"
            >
              <div
                class="googlesitekit-change-badge"
              >
                +18.5%
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-audience-segmentation-tile-metric"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-metric__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__container"
            >
              <div
                class="googlesitekit-audience-segmentation-tile-metric__value"
              >
                3
              </div>
              <div
                class="googlesitekit-audience-segmentation-tile-metric__title"
              >
                Visits per visitor
              </div>
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__badge-container"
            >
              <div
                class="googlesitekit-change-badge"
              >
                +50%
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-audience-segmentation-tile-metric"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-metric__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__container"
            >
              <div
                class="googlesitekit-audience-segmentation-tile-metric__value"
              >
                2
              </div>
              <div
                class="googlesitekit-audience-segmentation-tile-metric__title"
              >
                Pages per visit
              </div>
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__badge-container"
            >
              <div
                class="googlesitekit-change-badge googlesitekit-change-badge--negative"
              >
                -33.3%
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-audience-segmentation-tile-metric"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-metric__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__container"
            >
              <div
                class="googlesitekit-audience-segmentation-tile-metric__value"
              >
                1.6K
              </div>
              <div
                class="googlesitekit-audience-segmentation-tile-metric__title"
              >
                33.3% of total pageviews
              </div>
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__badge-container"
            >
              <div
                class="googlesitekit-change-badge"
              >
                +4.1%
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-audience-segmentation-tile-metric googlesitekit-audience-segmentation-tile-metric--cities"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-metric__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__container"
            >
              <div
                class="googlesitekit-audience-segmentation-tile-metric__title"
              >
                Cities with the most visitors
              </div>
              <div
                class="googlesitekit-audience-segmentation-tile-metric__content"
              >
                <div
                  class="googlesitekit-audience-segmentation-tile-metric__cities-metric"
                >
                  <div
                    class="googlesitekit-audience-segmentation-tile-metric__cities-metric-name"
                  >
                    Dublin
                  </div>
                  <div
                    class="googlesitekit-audience-segmentation-tile-metric__cities-metric-value"
                  >
                    63.8%
                  </div>
                </div>
                <div
                  class="googlesitekit-audience-segmentation-tile-metric__cities-metric"
                >
                  <div
                    class="googlesitekit-audience-segmentation-tile-metric__cities-metric-name"
                  >
                    London
                  </div>
                  <div
                    class="googlesitekit-audience-segmentation-tile-metric__cities-metric-value"
                  >
                    20.7%
                  </div>
                </div>
                <div
                  class="googlesitekit-audience-segmentation-tile-metric__cities-metric"
                >
                  <div
                    class="googlesitekit-audience-segmentation-tile-metric__cities-metric-name"
                  >
                    New York
                  </div>
                  <div
                    class="googlesitekit-audience-segmentation-tile-metric__cities-metric-value"
                  >
                    15.5%
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-audience-segmentation-tile-metric googlesitekit-audience-segmentation-tile-metric--top-content"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-metric__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-audience-segmentation-tile-metric__container"
            >
              <div
                class="googlesitekit-audience-segmentation-tile-metric__title"
              >
                Top content by pageviews
              </div>
              <div
                class="googlesitekit-audience-segmentation-tile-metric__content"
              >
                <div
                  class="googlesitekit-audience-segmentation-tile-metric__no-data"
                >
                  No data to show
                  <p>
                    Update Analytics to track metric
                  </p>
                  <button
                    class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised mdc-button--danger"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Update
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AudienceTile with zero data, in the partial data state should render the zero data tile 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--audienceTile googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-audience-segmentation-tile"
      >
        <div
          class="googlesitekit-audience-segmentation-tile__zero-data-container"
        >
          <div
            class="googlesitekit-audience-segmentation-tile__zero-data-content"
          >
            <svg />
            <p
              class="googlesitekit-audience-segmentation-tile__zero-data-description"
            >
              Site Kit is collecting data for this group.
            </p>
            <p
              class="googlesitekit-audience-segmentation-tile__zero-data-description"
            >
              You can hide this group until data is available.
            </p>
            <button
              class="googlesitekit-cta-link googlesitekit-audience-segmentation-tile-hide-cta googlesitekit-cta-link--secondary googlesitekit-cta-link--link-button"
            >
              <span
                class="googlesitekit-icon-wrapper"
                style="margin-right: 5px;"
              >
                <svg />
              </span>
              <span
                class="googlesitekit-cta-link__contents"
              >
                Temporarily hide
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
